### PR TITLE
Add frame rate limiting to GPU render

### LINKED ISSUE #1: Excessive GPU utilization
Currently the viewer greedily renders at unlimited framerate. This is not a problem for beefy systems, but it is taxing on modest streaming PCs which must also share GPU resources with video encoders for streaming.

### PR DESCRIPTION
**Pull request by ericjtaylor**
closes #1

### CODE REVIEW

File: include/echo_chan.hpp [hunk 1]
```diff
@@ -19,6 +19,8 @@ class EchoChan {
   void calculateRotation(uint8_t& state, uint8_t& prev_state, int& rgear);
   void extractStateFromPins();
   bool isButtonPressed(BUTTON_LABEL b);
```
Comment (HWXLR8) [inline include/echo_chan.hpp:21]: the whole point of getter and setter methods is to provide a nice interface to the underlying state uncoupled from the implementation. the practice of declaring some private variable then providing trivial `getVariable()` and `setVariable(var x)` seems like complexity for the sake of complexity. just modify the variable directly at that point and avoid the function call. petty rant aside, as I foresee this state tracking may get more complicated in the future, I vote for keeping it in, but instead providing an entirely decoupled interface:

`void setStateFresh(bool fresh);`

changed to

`void setFreshState();`
`void setStaleState();`

and  `bool getStateFresh();` -> `bool isStateFresh()`

Comment (HWXLR8) [inline include/echo_chan.hpp:21]: a separate concern I have is whether this even belongs in the EchoChan class. the echo chan class is strictly for interfacing with the echo chan hardware, where as this fresh state tracking seems only concerned with the input viewer update mechanism. you did not provide any description for this PR, so I can only guess what your intentions are here.

**Review by HWXLR8 — CHANGES_REQUESTED** [review]
please give me some explanation as to what exactly was done here. as it stands now, I am not convinced your changes belong in the EchoChan class as they seem more related to the input viewer. there also seems to be changes that have nothing to do with the rate limiting mixed in.

File: include/echo_chan.hpp [hunk 2]
```diff
@@ -19,6 +19,8 @@ class EchoChan {
...
+  bool getStateFresh();
+  void setStateFresh(bool fresh);
```
Comment (HWXLR8) [inline include/echo_chan.hpp:23]: the whole point of getter and setter methods is to provide a nice interface to the underlying state uncoupled from the implementation. the practice of declaring some private variable then providing trivial `getVariable()` and `setVariable(var x)` seems like complexity for the sake of complexity. just modify the variable directly at that point and avoid the function call. petty rant aside, as I foresee this state tracking may get more complicated in the future, I vote for keeping it in, but instead providing an entirely decoupled interface:

`void setStateFresh(bool fresh);`

changed to

`void setFreshState();`
`void setStaleState();`

and  `bool getStateFresh();` -> `bool isStateFresh()`

Comment (HWXLR8) [inline include/echo_chan.hpp:23]: a separate concern I have is whether this even belongs in the EchoChan class. the echo chan class is strictly for interfacing with the echo chan hardware, where as this fresh state tracking seems only concerned with the input viewer update mechanism. you did not provide any description for this PR, so I can only guess what your intentions are here.

File: include/echo_chan.hpp [hunk 3]
```diff
@@ -31,4 +33,5 @@ class EchoChan {
   double r1_;
   double r2_;
   bool running_ = false;
+  bool fresh_ = true;
```
Comment (HWXLR8) [inline include/echo_chan.hpp:36]: it is not clear what the word "fresh" means in this context. the name tells me that there must exist "fresh" and "stale" states, but tells me nothing about those states? what counts as fresh and what counts as stale? I would prefer to use a descriptive name such as "has_data_been_read" (if that is even what you are trying to track here).

File: src/echo_chan.cpp [hunk 4]
```diff
@@ -1,28 +1,27 @@
...
-    port = "COM" + std::to_string(i);
+    std::string port = "COM" + std::to_string(i);
     char byte[1] = { 0 };
-    std::cout << "ATTEMPTING TO CONNECT TO " << port << std::endl;
 
     // attempt to open the device
     serial_ = CreateFile(port.c_str(), GENERIC_READ | GENERIC_WRITE, 0, 0, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, 0);
     if (serial_ != INVALID_HANDLE_VALUE) {
+      std::cout << "Attempting handshake with " << port << std::endl;
+
```
Comment (HWXLR8) [inline src/echo_chan.cpp:24]: these changes are unrelated to frame rate limiting and make it difficult to understand what is and is not related to your changes, please break these into 2 separate commits: one for cleanup, and one for rate limiting.

File: src/echo_chan.cpp [hunk 5]
```diff
@@ -51,19 +50,10 @@ EchoChan::EchoChan() {
...
 
-  // set generous (0 = unlimited) timeouts now that we know it's Echo Chan
-  COMMTIMEOUTS timeout = { 0 };
-  timeout.ReadIntervalTimeout = 0;
-  timeout.ReadTotalTimeoutConstant = 0;
-  timeout.ReadTotalTimeoutMultiplier = 0;
-  timeout.WriteTotalTimeoutConstant = 50;
-  timeout.WriteTotalTimeoutMultiplier = 1;
-  SetCommTimeouts(serial_, &timeout);
-
```
Comment (HWXLR8) [inline src/echo_chan.cpp:62]: why was the timeout removed?

**Review by HWXLR8 — CHANGES_REQUESTED** [review]

File: include/echo_chan.hpp (hunk 2 above)
Comment (ericjtaylor) [reply include/echo_chan.hpp:23]: The complication is that a render may have `fresh` work initiated from either side. New Echo Chan inputs, or new GUI changes. Would it be better with 2 `fresh` variables, one in each class, with the render loop checking both each time?

Comment (ericjtaylor) [reply include/echo_chan.hpp:23]: Actually saying this out loud it sounds like it should live in GLPRG class...
